### PR TITLE
Implementation erc20

### PR DESCRIPTION
**Pull request by Cheelax**
Draft Implementation of ERC20 in cairo1.
Components done, systems todo

### CODE REVIEW

File: lib/src/erc20/erc20_base_subsystem.cairo [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    #[external]
+    fn transfer(recipient: ContractAddress, amount: u256) {
+        let sender = get_caller_address();
+        transfer_helper(sender, recipient, amount);
+    }
+
+    #[external]
+    fn transfer_from(sender: ContractAddress, recipient: ContractAddress, amount: u256) {
```
Comment (tarrencev) [inline lib/src/erc20/erc20_base_subsystem.cairo:80]: The logic for  `transfer_from` and `approve` need to be moved to systems.

For example approval could look something like this:

```
#[system]
mod ERC20_Approve {
    execute(token_id: felt252, sender: ContractAddress, recipient: ContractAddress, amount: u256) {
        let approval_sk: StorageKey = (token_id, (sender.into()).into();
        let approval = commands::<Approval>::get(approval_sk);
        commands::set(approval_sk, (
            Approval { amount: approval.amount + amount }
        ))
    }
}
```

File: lib/src/erc20/erc20_base_subsystem.cairo [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        let caller = get_caller_address();
+        spend_allowance(sender, caller, amount);
+        transfer_helper(sender, recipient, amount);
+    }
+
+    //approval system
+    #[external]
+    fn approve(spender: ContractAddress, amount: u256) {
```
Comment (tarrencev) [inline lib/src/erc20/erc20_base_subsystem.cairo:88]: The logic for  `transfer_from` and `approve` need to be moved to systems.

For example approval could look something like this:

```
#[system]
mod ERC20_Approve {
    execute(token_id: felt252, spender: ContractAddress, amount: u256) {
        let caller = get_caller_address();
        let approval_sk: StorageKey = (token_id, (caller, spender)).into();
        let approval = commands::<Approval>::get(approval_sk);
        commands::set(approval_sk, (
            Approval { amount: approval.amount + amount }
        ))
    }
}
```

File: lib/src/erc20/erc20_base_subsystem.cairo (hunk 1 above)
Comment (Cheelax) [reply lib/src/erc20/erc20_base_subsystem.cairo:80]: Do we agree that I need to move other methods than the two above too? In two files one for approve one for balance? Or one file for each method?

Comment (tarrencev) [reply lib/src/erc20/erc20_base_subsystem.cairo:80]: Yes i think so. Each file can have one system

File: lib/src/erc20/components/ownership_component.cairo [hunk 3]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (tarrencev) [inline lib/src/erc20/components/ownership_component.cairo:1]: we can remove the `_component` from the names here

File: lib/src/erc20/erc20_base_subsystem.cairo [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    use starknet::ContractAddress;
+    use starknet::ContractAddressZeroable;
+
+    struct Storage {
+        name: felt252,
+        symbol: felt252,
+        decimals: u8,
+        total_supply: u256,
+        balances: OwnershipComponent,
+        allowances: ApprovalComponent,
```
Comment (tarrencev) [inline lib/src/erc20/erc20_base_subsystem.cairo:17]:
```suggestion
```

Component state is stored in the world

File: lib/src/erc20/systems/approval/erc20_approve.cairo [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#[system]
```
Comment (tarrencev) [inline lib/src/erc20/systems/approval/erc20_approve.cairo:1]: we only need the approve system. increase / decrease can be implemented in the base and use it

File: lib/src/erc20/systems/approval/erc20_decrease_allowance.cairo [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+mod ERC20_Decrease_Allowance {
+    execute(token_id: felt252, spender: ContractAddress, substracted_value: u256) {
+        let caller = get_caller_address();
+        let approval_sk: StorageKey = (token_id, (caller, spender)).into();
+        let approval = commands::<Approval>::get(approval_sk);
+        commands::set(approval_sk, (
+            Approval { amount: approval.amount - substracted_value }
+        ))
+    }
+}
```
Comment (tarrencev) [inline lib/src/erc20/systems/approval/erc20_decrease_allowance.cairo:11]:
```suggestion
```

File: lib/src/erc20/systems/approval/erc20_increase_allowance.cairo [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+mod ERC20_Increase_Allowance {
+    execute(token_id: felt252, spender: ContractAddress, added_value: u256) {
+        let caller = get_caller_address();
+        let approval_sk: StorageKey = (token_id, (caller, spender)).into();
+        let approval = commands::<Approval>::get(approval_sk);
+        commands::set(approval_sk, (
+            Approval { amount: approval.amount + added_value }
+        ))
+    }
+}
```
Comment (tarrencev) [inline lib/src/erc20/systems/approval/erc20_increase_allowance.cairo:11]:
```suggestion
```

File: lib/src/erc20/systems/approval/erc20_approve.cairo [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+mod ERC20_Approve {
+    execute(token_id: felt252, spender: ContractAddress, amount: u256) {
+        let caller = get_caller_address();
+        let approval_sk: StorageKey = (token_id, (caller, spender)).into();
```
Comment (tarrencev) [inline lib/src/erc20/systems/approval/erc20_approve.cairo:5]:
```suggestion
        let approval_sk: StorageKey = (token_id, (caller.into(), spender)).into();
```

you'll need to convert `ContractAddress` into a `felt252`. You might have to import the trait too

File: lib/src/erc20/systems/approval/erc20_approve.cairo [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+        let approval = commands::<Approval>::get(approval_sk);
+        commands::set(approval_sk, (
+            Approval { amount: approval.amount + amount }
```
Comment (tarrencev) [inline lib/src/erc20/systems/approval/erc20_approve.cairo:8]:
```suggestion
            Approval { amount: amount }
```

File: lib/src/erc20/systems/transfer/erc20_transfer.cairo [hunk 10]
```diff
@@ -0,0 +1,24 @@
+#[system]
+mod ERC20_Transfer {
```
Comment (tarrencev) [inline lib/src/erc20/systems/transfer/erc20_transfer.cairo:2]: lets move this and approve into the parent folder

File: lib/src/erc20/systems/transfer/erc20_transfer.cairo [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    execute(token_id: felt252, recipient: ContractAddress, value: u256) {
```
Comment (tarrencev) [inline lib/src/erc20/systems/transfer/erc20_transfer.cairo:3]: i think we want this to implement `transfer_from`, so we can 
```suggestion
    execute(token_id: felt252, spender: ContractAddress, recipient: ContractAddress, amount: felt) {
```

File: lib/src/erc20/systems/transfer/erc20_transfer.cairo [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+        let caller = get_caller_address();
```
Comment (tarrencev) [inline lib/src/erc20/systems/transfer/erc20_transfer.cairo:4]:
```suggestion
```

File: lib/src/erc20/systems/transfer/erc20_transfer.cairo [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        assert(!sender.is_zero(), 'ERC20: transfer from 0');
+        assert(!recipient.is_zero(), 'ERC20: transfer to 0');
+
+        let ownership_sender_sk: StorageKey = (token_id, (caller, spender)).into();
```
Comment (tarrencev) [inline lib/src/erc20/systems/transfer/erc20_transfer.cairo:8]:
```suggestion
        let ownership_spender_sk: StorageKey = (token_id, (spender)).into();
```

we want the key here to be the spender address only

File: lib/src/erc20/systems/transfer/erc20_transfer.cairo [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+        let ownership_receiver_sk: StorageKey = (token_id, (caller, spender)).into();
```
Comment (tarrencev) [inline lib/src/erc20/systems/transfer/erc20_transfer.cairo:9]:
```suggestion
        let ownership_recipient_sk: StorageKey = (token_id, (recipient)).into();
```

File: lib/src/erc20/systems/transfer/erc20_transfer.cairo [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+        let ownership = commands::<Ownership>::get(ownership_sk);
+        //add storage
+
+        commands::set(ownership_sender_sk, (
+            Ownership { balance : ownership.balance - amount}
+        ));
+
+        commands::set(ownership_receiver_sk, (
+            Ownership { balance : ownership.balance + amount}
+        ));
```
Comment (tarrencev) [inline lib/src/erc20/systems/transfer/erc20_transfer.cairo:19]:
```suggestion
        let ownership_spender = commands::<Ownership>::get(ownership_spender_sk);
        commands::set(ownership_spender_sk, (
            Ownership { balance : ownership_spender.balance - amount}
        ));

        let receipient_ownership = = commands::<Ownership>::get(ownership_recipient_sk);
        commands::set(ownership_recipient_sk, (
            Ownership { balance : ownership.balance + amount}
        ));
```

File: lib/src/erc20/systems/transfer/erc20_transfer.cairo [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+        //this is the event
+        Transfer(sender, recipient, amount);
```
Comment (tarrencev) [inline lib/src/erc20/systems/transfer/erc20_transfer.cairo:22]:
```suggestion
```

The event should be emitted from the base contract

File: lib/src/erc20/systems/transfer/erc20_transfer_from.cairo [hunk 17]
```diff
@@ -0,0 +1,3 @@
+#[system]
+mod ERC20_Transfer_From {
+}
```
Comment (tarrencev) [inline lib/src/erc20/systems/transfer/erc20_transfer_from.cairo:3]:
```suggestion
```

**Review by tarrencev — CHANGES_REQUESTED** [review]

File: lib/src/erc20/systems/transfer/erc20_transfer.cairo (hunk 16 above)
Comment (Cheelax) [reply lib/src/erc20/systems/transfer/erc20_transfer.cairo:22]: I put the event in erc_base_subsystem, not sure if it was the expected

File: lib/src/erc20/erc20_base_subsystem.cairo (hunk 4 above)
Comment (Cheelax) [reply lib/src/erc20/erc20_base_subsystem.cairo:17]: I'm not sure of how to put the "contract" erc20 and Components working together in the world, would like to see it with you

**Comment by Cheelax** [comment]
Any input on what's next regarding this feature?

**Pull request merged**